hid-xpadneo/src/hid-xpadneo.c

```diff
@@ -899,7 +906,12 @@ static int xpadneo_raw_event(struct hid_device *hdev, struct hid_report *report,
 		bits |= (data[14] & (BIT(0) | BIT(1))) >> 0;	/* A, B */
 		bits |= (data[14] & (BIT(3) | BIT(4))) >> 1;	/* X, Y */
 		bits |= (data[14] & (BIT(6) | BIT(7))) >> 2;	/* LB, RB */
-		bits |= (data[16] & BIT(0)) << 6;	/* Back */
+		if (xdata->quirks & XPADNEO_QUIRK_SHARE_BUTTON) {
+			/* correct button mapping of Xbox controllers with a share button */
+			bits |= (data[15] & BIT(2)) ? BIT(6) : 0; /* Back */
```

Comment:
Looks complicated... Please try to avoid the ternary operator to reflect how the other lines do it. It took me some minutes to realize what this says. ;-) It's probably more easy to read (put bit 6 if bit 2 is set) but the other lines shift bits instead of explicitly setting them.

Reply:
I found it the opposite way when reading it: the new position is dependent on the old one, and I had to do math in my head to figure out where that was. This was just my first get-it-working pass, so I didn't bother to calculate the bitshift. I'll do that though tomorrow.

Reply:
Yeah, I figured that out eventually. I came from the other side when I developed this piece of code because we had a bit mask of
```
...11011011
...RLZYXCBA
```
and I've just gone: "okay, take these two, skip 1, take the next 2, skip 1 - ah okay, that shift 1, then shift 2..."

Also, I think the resulting assembler code is better optimized without using ternary operators. So I decided to continue that way even if that meant to do some math in the head. :-)

Sleep well...

Reply:
To be fair, ternaries [generate](https://godbolt.org/z/Pjc1no) fewer instructions without optimization and identical instructions with even `-O1` :smile:.

Reply:
Oh, I didn't know... Then maybe it's time to rewrite all the lines. It's up to you if you want to do it or if I do it.

Reply:
Cool site, btw... When I researched that, people always insisted that ternaries generate additional code overhead that cannot be properly optimized, and that it generates jumps the hurts the branch prediction performance... Well, looks like things may actually be different. Tho, your example is optimized for an immediate return and could also be inlined by the compiler. I'm not sure what the cost of a failed branch prediction is. Also, that example evaluates statically: In the O1 case, both are identical.